Deploy FreeTDS as AWS Lambda Layer

BrianBorge opened the pull request:
In order to use the TinyTDS gem with ruby AWS Lambda Functions, we need to have access to FreeTDS binaries that AWS does not provide with its execution environment.

The solution is to use AWS Lambda Layers [1].

The provided deploy script builds FreeTDS on architecture that runs on AWS Lambda and, using the AWS CLI, publishes the artifact as an AWS Lambda layer for functions to use.


1: https://docs.aws.amazon.com/lambda/latest/dg/configuration-layers.html

BrianBorge committed:
Deploy FreeTDS as a AWS Lambda Layer for use with AWS Lambda Functions

aharpervc commented on aws-lambda-layer/build.sh:
```diff
@@ -0,0 +1,18 @@
+set -e
+
+FREETDS_VERSION=${FREETDS_VERSION:=1.1.6}
+IMAGE_NAME=veracross/freetds-layer:${FREETDS_VERSION:latest}
```
`veracross/freetds`

BrianBorge replied:
How do we want to tag these images?

`veracross/freetds:aws-<freetds-version>`?

aharpervc commented on aws-lambda-layer/build.sh:
```diff
@@ -0,0 +1,18 @@
+rm -rf ./build
+mkdir ./build
+
+docker build \
```
should there also be a `docker push` step in here?

BrianBorge replied:
Probably - I'll add the step.

aharpervc commented on aws-lambda-layer/Dockerfile:
```diff
@@ -0,0 +1,26 @@
+
+RUN mkdir $INSTALLDIR
+
+RUN wget http://www.freetds.org/files/stable/freetds-${FREETDS_VERSION}.tar.gz && \
```
probably want to do this: https://github.com/veracross/freetds/blob/master/Dockerfile#L6

aharpervc reviewed (commented):
See comments

aharpervc commented on aws-lambda-layer/Dockerfile:
```diff
@@ -0,0 +1,25 @@
+
+RUN mkdir $INSTALLDIR
+
+RUN wget https://github.com/veracross/freetds/raw/master/freetds/freetds-${FREETDS_VERSION}.tar.gz && \    tar -xzf freetds-${FREETDS_VERSION}.tar.gz && \
```
should this be broken up to several lines, like the other commands are?

BrianBorge replied:
Yep, it was a typo.

aharpervc commented on aws-lambda-layer/README.md:
```diff
@@ -0,0 +1,48 @@
+
+The solution is to use AWS Lambda Layers.
+
+The provided deploy script builds FreeTDS on architecture that runs in AWS Lambda and, using the AWS CLI, publishes the artifact as an AWS Lambda layer for functions to use.
```
"on the same architecture"?

aharpervc commented on aws-lambda-layer/deploy.sh:
```diff
@@ -0,0 +1,13 @@
+
+FREETDS_VERSION=${FREETDS_VERSION:=1.1.6}
+
+./build.sh
```
How do you feel about making a Rakefile, and doing these as rake tasks instead of shell scripts? There's a couple benefits, such as a central place to list out all the work, and also you can chain dependencies instead of running another script inside the first script like this.

BrianBorge replied:
I like that idea.

BrianBorge replied:
Question...where should the Rakefile live?  In the root of the project or the subdirectory, `aws-lambda-layer/`?

If the Rakefile live in the root, then the tasks should be namespaced probably to "layer:build" and "layer:deploy"

aharpervc replied:
Hm, that's interesting. I'm not opposed, although I think for now lets just prototype it for this AWS work and keep it in this folder

aharpervc commented on aws-lambda-layer/README.md:
````diff
@@ -0,0 +1,48 @@
+```
+
+## Additional Resources
+- https://aws.amazon.com/cli/
````
Line break after header?

smiller-vc commented on aws-lambda-layer/Rakefile:
```diff
@@ -0,0 +1,34 @@
+
+BUILD_DIR = File.join(__dir__, 'build')
+
+ENV['FREETDS_VERSION'] ||= '1.1.6'
```
Is 1.1.6 the latest version? Struggling any reference to FreeTDS editions online.

BrianBorge replied:
1.1.40 is the latest release.  I defaulted to 1.1.6 as it was the latest at the time of writing.  I can add a link to where to find the latest release.  Although the way these tasks are written, we are pulling freetds from this repo since their ftp goes down all the time so you're limited to the versions of freetds we have added to this repo which at the time of this writing are 1.00.110 and 1.1.6.

1:  https://www.freetds.org/software.html

smiller-vc replied:
I definitely did not read "1.1.40" on the main site as "1.1.4" and think we were on some experimental release...nope would never do that...

![naw](https://media.giphy.com/media/dCixq2fQBT2xy/giphy.gif)

BrianBorge replied:
I was confused for a solid minute before linking you the site as well :)

smiller-vc commented on aws-lambda-layer/build.sh:
```diff
@@ -0,0 +1,20 @@
+#!/bin/bash -x
```
Is this file now irrelevant with the introduction of the Rakefile?

BrianBorge replied:
Not anymore, I deleted it.

smiller-vc commented on aws-lambda-layer/deploy.sh:
```diff
@@ -0,0 +1,13 @@
+
+FREETDS_VERSION=${FREETDS_VERSION:=1.1.6}
+
+./build.sh
```
Should this now point to the Rakefile?

BrianBorge replied:
With these tasks moved to a Rakefile, this file will be deleted.

smiller-vc commented on aws-lambda-layer/README.md:
```diff
@@ -0,0 +1,48 @@
+
+## Usage
+
+- Run `./deploy.sh`
```
Is this now the Rakefile too?

BrianBorge replied:
yep, changed to run `rake deploy`

smiller-vc reviewed (commented):
I realize I'm leaving a review in the middle of Andrew's review, so just some small questions you're probably already working on (and for my own learning).

Also - so I'm clear on the process here - rather than building all of this in an EC2 instance, [as in the documentation](https://aws.amazon.com/premiumsupport/knowledge-center/lambda-linux-binary-package/), we're executing this build in a Docker container, then deploying it as a Lambda layer, correct?

BrianBorge commented:
> Also - so I'm clear on the process here - rather than building all of this in an EC2 instance, as in the documentation, we're executing this build in a Docker container, then deploying it as a Lambda layer, correct?

That's correct, we can use a docker image that replicates the lambda environment and runtime we are using (ruby2.5) without needing to launch and use an EC2 instance.

1:  https://github.com/lambci/docker-lambda

BrianBorge commented:
@aharpervc, @smiller-vc,

I've addressed the review comments.  The biggest change was moving the work done by the shell scripts into rake tasks that can be listed out with `rake -T` and can depend on each other (build task is executed before the deploy task).

Although not necessary, we could consider using a [ruby docker client](https://github.com/swipely/docker-api) instead of making system calls to Docker in the Rakefile.

Note: I haven't looked at the status/condition/etc of that ruby gem I linked.  It is just an idea to consider.  I don't think we need to take a dependency on a ruby docker client here.

BrianBorge commented:
We should also consider having a DevOps team member review this.  Any team that wants to build a lambda function that depends on FreeTDS can make use of the deployed layers. I'll tag their team.

aharpervc commented:
> Although not necessary, we could consider using a [ruby docker client](https://github.com/swipely/docker-api) instead of making system calls to Docker in the Rakefile.
> 
> Note: I haven't looked at the status/condition/etc of that ruby gem I linked. It is just an idea to consider. I don't think we need to take a dependency on a ruby docker client here.

Yeah I did a survey of libraries when I was adding container building support to vcli, and there was some reason that it wasn't going to do what I needed. Can't recall precisely now, and maybe whatever it is has been fixed.

aharpervc commented on aws-lambda-layer/Rakefile:
```diff
@@ -0,0 +1,47 @@
+
+desc 'Build FreeTDS AWS Lambda Layer'
+task :build, [:freetds_version] do |_t, args|
+  version = args[:freetds_version] || ENV['FREETDS_VERSION'] || '1.1.6'
```
can you pull `1.1.6` out to a top level constant?

BrianBorge replied:
Sure thing.

aharpervc reviewed (approved):
If it works, LGTM. While Devops input is always great to have, I think in this case the PR can probably be merged if it works for us regardless. We can always revisit in the future if there's something that needs to change. (An example of something that might change is if Devops provides a new base image for us that includes freetds, and we can sidestep all of this)

aharpervc commented on aws-lambda-layer/README.md:
```diff
@@ -0,0 +1,44 @@
+
+## Requirements
+
+- Ruby 2.5.3
```
Is this rake task really that specific?

BrianBorge replied:
No, I guess not but you do need to ruby to run the rake tasks and the layer is built on a container using ruby2.5 so I figure I would add 

This line could just be "Ruby 2.5.x" or just "Ruby".

aharpervc commented on aws-lambda-layer/README.md:
```diff
@@ -0,0 +1,44 @@
+
+- Ruby 2.5.3
+- Docker
+- AWS CLI
```
This is only for deploys to real infrastructure, right? If I'm building locally, do I need this?

BrianBorge replied:
If you're building locally you need ruby and docker.  If you're deploying a function to production then you need the aws cli.

aharpervc commented on aws-lambda-layer/README.md:
```diff
@@ -0,0 +1,44 @@
+
+## Usage
+
+- Run `rake deploy`
```
Should there be an example for what to do when all you want is to work on a lambda function locally?

BrianBorge replied:
No, I don't think so.  If you want to work on a lambda function locally you technically don't need this layer.  It's only important when deploying to AWS.

When you want to build a function locally and test it out, we make use of a custom image that has the binaries built into it already which we have called "veracross/api-build-environment" which highjacks the SAM CLI build step.

smiller-vc commented:
> That's correct, we can use a docker image that replicates the lambda environment and runtime we are using (ruby2.5) without needing to launch and use an EC2 instance.
> 
> 1: https://github.com/lambci/docker-lambda

You should totally write a blog post about that, that's a great idea! (And so we can save the next poor soul that attempts this some trouble)
